corehq/apps/enterprise/enterprise.py

```diff
@@ -405,6 +416,92 @@ def _get_individual_export_rows(self, exports, export_line_counts):
         return rows
 
 
+class EnterpriseCommCareVersionReport(EnterpriseReport):
+    title = gettext_lazy('CommCare Version Compliance')
+    total_description = gettext_lazy('% of Mobile Workers on the Latest CommCare Version')
+
+    @property
+    def headers(self):
+        return [
+            _('Mobile Worker'),
+            _('Project Space'),
+            _('Latest Version Available at Submission'),
+            _('Version in Use'),
+        ]
+
+    @property
+    def rows(self):
+        rows = []
+        config = CommCareBuildConfig.fetch()
+        version_cache = {}
+        for domain in self.account.get_domains():
+            rows.extend(self.rows_for_domain(domain, config, version_cache))
+        return rows
+
+    @property
+    def total(self):
+        total_mobile_workers = 0
+        total_up_to_date = 0
+        config = CommCareBuildConfig.fetch()
+        version_cache = {}
+
+        def total_for_domain(domain):
+            mobile_workers = get_mobile_user_count(domain, include_inactive=False)
+            if mobile_workers == 0:
+                return 0, 0
+            outdated_users = len(self.rows_for_domain(domain, config, version_cache))
+            return mobile_workers, outdated_users
```

Comment:
I feel like nested functions generally make the code more confusing. One good use for inner functions is when the inner functions depends on a value from the outer function (i.e. closures). A passable reason would be if the function is extremely simple (a one-liner) and creating an inner function improves readability. I don't think this function meets either of those criteria, though. Nesting it this way means it becomes harder to debug the nested function.

Reply:
Oh...I guess it technically is a closure, as it uses the `version_cache`. I guess the choice to nest this function was to avoid having `total_for_domain` take the `version_cache` parameter?

Reply:
My initial intention is not avoid having total_for_domain take the version_cache parameter. But I think you just find another reason to next it. :joy: I nest it in total because, otherwise total_for_domain looks like I want to override the total_for_domain in the base class. While I have no intention to override it, I will compute total in a special way in this tile.